README.txt
=== Connections Business Directory Certifications ===
Contributors: shazahm1@hotmail.com
Donate link: https://connections-pro.com/
Tags: business directory, certifications, directory
Requires at least: 4.7.12
Tested up to: 5.4
Requires PHP: 5.6.20
Stable tag: 1.4
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

An extension for the Connections Business Directory plugin which adds the ability to add and assign certifications to your business directory entries.

== Description ==

This is an extension plugin for the [Connections Business Directory plugin](http://wordpress.org/plugins/connections/) please be sure to install and active it before adding this plugin.

What does this plugin do? It adds the ability to add certifications and assign them to you business directory entries.

The certifications of an entry can optionally be shown as a content block within a template or with the supplied widget int the theme's sidebar.

= Extensible =

Here are some great **free extensions** (with more on the way) that enhance your experience with Connections Business Directory:

**Utility**

* [Toolbar](https://wordpress.org/plugins/connections-toolbar/) :: Provides quick links to the admin pages from the admin bar.
* [Login](https://wordpress.org/plugins/connections-business-directory-login/) :: Provides a simple to use login shortcode and widget.
* [Anniversary and Birthday Emails](https://wordpress.org/plugins/connections-business-directory-anniversary-and-birthday-emails) :: Add the ability to automatically send a customizable email to entries on their anniversary or birthday.

**Custom Fields**

* [Business Open Hours](https://wordpress.org/plugins/connections-business-directory-hours/) :: Add the business open hours.
* [Local Time](https://wordpress.org/plugins/connections-business-directory-local-time/) :: Add the business local time.
* [Facilities](https://wordpress.org/plugins/connections-business-directory-facilities/) :: Add the business facilities.
* [Income Level](https://wordpress.org/plugins/connections-business-directory-income-levels/) :: Add an income level.
* [Education Level](https://wordpress.org/plugins/connections-business-directory-education-levels/) :: Add an education level.
* [Languages](https://wordpress.org/plugins/connections-business-directory-languages/) :: Add languages spoken.
* [Hobbies](https://wordpress.org/plugins/connections-business-directory-hobbies/) :: Add hobbies.

**Misc**

* [Face Detect](https://wordpress.org/plugins/connections-business-directory-face-detect/) :: Applies face detection before cropping an image.

**[Premium Extensions](https://connections-pro.com/extensions/)**

* [Authored](https://connections-pro.com/add-on/authored/) :: Displays a list of blog posts written by the entry on their profile page.
* [Contact](https://connections-pro.com/add-on/contact/) :: Displays a contact form on the entry's profile page to allow your visitors to contact the entry without revealing their email address.
* [CSV Import](https://connections-pro.com/add-on/csv-import/) :: Bulk import your data in to your directory.
* [Custom Category Order](https://connections-pro.com/add-on/custom-category-order/) :: Order your categories exactly as you need them.
* [Custom Entry Order](https://connections-pro.com/add-on/custom-entry-order/) :: Allows you to easily define the order that your business directory entries should be displayed.
* [Enhanced Categories](https://connections-pro.com/add-on/enhanced-categories/) :: Adds many features to the categories.
* [Form](https://connections-pro.com/add-on/form/) :: Allow site visitor to submit entries to your directory. Also provides frontend editing support.
* [Link](https://connections-pro.com/add-on/link/) :: Links a WordPress user to an entry so that user can maintain their entry with or without moderation.
* [ROT13 Encryption](https://connections-pro.com/add-on/rot13-email-encryption/) :: Protect email addresses from being harvested from your business directory by spam bots.
* [SiteShot](https://connections-pro.com/add-on/siteshot/) :: Show a screen capture of the entry's website.
* [Widget Pack](https://connections-pro.com/add-on/widget-pack/) :: A set of feature rich, versatile and highly configurable widgets that can be used to enhance your directory.

**[Premium Templates](https://connections-pro.com/templates/)**

Connections Business Directory comes with a couple templates to get you started which fully support the [Template Customizer](https://connections-pro.com/2015/07/27/feature-preview-template-customizer/). In addition to these free templates there are many premium templates available to take your business directory the level visually.

* [Circled](https://connections-pro.com/add-on/circled/) :: A simple but bold template specifically designed for displaying small team or staff directory.
* [cMap](https://connections-pro.com/add-on/cmap/) :: Our most popular go to template for a business directory and chamber of commerce business directory, featuring full Template Customizer support.
* [Excerpt Plus](https://connections-pro.com/add-on/excerpt-plus/) :: Simple design which features a directory entry name, photo thumbnail and excerpt.
* [Gridder](https://connections-pro.com/add-on/gridder/) :: Another simple but bold template which displays the directory in a grid layout. Perfect for displaying a small team or staff directory
* [Slim Plus](https://connections-pro.com/add-on/slim-plus/) :: Specifically designed to take up as little space on the page as possible.
* [Tile Plus](https://connections-pro.com/add-on/tile-plus/) :: This template was purposely designed to display the directory in a grid or column layout.

== Installation ==

= Using the WordPress Plugin Search =

1. Navigate to the `Add New` sub-page under the Plugins admin page.
2. Search for `connections business directory certifications `.
3. The plugin should be listed first in the search results.
4. Click the `Install Now` link.
5. Lastly click the `Activate Plugin` link to activate the plugin.

= Uploading in WordPress Admin =

1. [Download the plugin zip file](https://wordpress.org/plugins/connections-business-directory-certifications/) and save it to your computer.
2. Navigate to the `Add New` sub-page under the Plugins admin page.
3. Click the `Upload` link.
4. Select Connections Business Directory Certifications zip file from where you saved the zip file on your computer.
5. Click the `Install Now` button.
6. Lastly click the `Activate Plugin` link to activate the plugin.

= Using FTP =

1. [Download the plugin zip file](https://wordpress.org/plugins/connections-business-directory-certifications/) and save it to your computer.
2. Extract the Connections Business Directory Certifications zip file.
3. Create a new directory named `connections-business-directory-certifications ` directory in the `../wp-content/plugins/` directory.
4. Upload the files from the folder extracted in Step 2.
4. Activate the plugin on the Plugins admin page.

== Frequently Asked Questions ==

None yet...

== Screenshots ==

1. Add new business certifications admin page.
2. Assign certifications to business directory listing.
3. Certification Content Block.
4. Certification widget display.

== Changelog ==

= 1.4 05/13/2020 =
* I18N: Add German (Germany) translation.
* OTHER: Update plugin naming to match current convention.
* OTHER: Update copyright year.
* OTHER: Update plugin tags.
* OTHER: Bump "Tested up to" to version 5.4.
* OTHER: Bump "Requires PHP" to version 5.6.20.
* OTHER: Bump "Requires at least" to version 4.7.12.
* OTHER: Update URLs to `https`.

= 1.3 08/01/2018 =
* FEATURE: Add support for the Gravity Forms Connector.
* I18N: Update Dutch (Netherlands) translation files. Props Age!

= 1.2 08/01/2018 =
* FEATURE: Add support for CSV export.
* BUG: Correct fatal bug. Calling nonexistent method.

= 1.1 07/19/2018 =
* FEATURE: Add support for the [CSV Import extension](https://connections-pro.com/add-on/csv-import/) so certifications can now be imported and attached to entries.
* TWEAK: Utilize `cnText_Domain` to load translations.
* I18N: Add POT file.
* I18N: Add Dutch (Netherlands) translation. props Age
* OTHER: Update readme.txt.
* OTHER: Update plugin header copyright year.
* OTHER: Update plugin tags.
* OTHER: Bump requires WP version to 4.5.3 to match requirements of core Connections plugin.

= 1.0 10/26/2017 =
* Initial release.

== Upgrade Notice ==

= 1.0 =
Initial release.

= 1.1 =
It is recommended to backup before updating. Requires WordPress >= 4.5.3 and PHP >= 5.3 PHP version >= 7.1 recommended.

= 1.2 =
It is recommended to backup before updating. Requires WordPress >= 4.5.3 and PHP >= 5.4 PHP version >= 7.1 recommended.

= 1.3 =
It is recommended to backup before updating. Requires WordPress >= 4.5.3 and PHP >= 5.4 PHP version >= 7.1 recommended.

= 1.4 =
It is recommended to backup before updating. Requires WordPress >= 4.7.12 and PHP >= 5.6.20 PHP version >= 7.1 recommended.


CHANGELOG.txt
== Changelog ==

= 1.4 05/13/2020 =
* I18N: Add German (Germany) translation.
* OTHER: Update plugin naming to match current convention.
* OTHER: Update copyright year.
* OTHER: Update plugin tags.
* OTHER: Bump "Tested up to" to version 5.4.
* OTHER: Bump "Requires PHP" to version 5.6.20.
* OTHER: Bump "Requires at least" to version 4.7.12.
* OTHER: Update URLs to `https`.

= 1.3 08/01/2018 =
* FEATURE: Add support for the Gravity Forms Connector.
* I18N: Update Dutch (Netherlands) translation files. Props Age!

= 1.2 08/01/2018 =
* FEATURE: Add support for CSV export.
* BUG: Correct fatal bug. Calling nonexistent method.

= 1.1 07/19/2018 =
* FEATURE: Add support for the CSV Import extension so certifications can now be imported and attached to entries.
* TWEAK: Utilize `cnText_Domain` to load translations.
* I18N: Add POT file.
* I18N: Add Dutch (Netherlands) translation. props Age
* OTHER: Update readme.txt.
* OTHER: Update plugin header copyright year.
* OTHER: Update plugin tags.
* OTHER: Bump requires WP version to 4.5.3 to match requirements of core Connections plugin.

= 1.0 03/09/2018 =
* Initial commit.


includes/class.widgets.php
<?php
/**
 * Static class for registering and displaying the widget.
 *
 * @package     Connections Certifications
 * @subpackage  Widget
 * @copyright   Copyright (c) 2017, Steven A. Zahm
 * @license     http://opensource.org/licenses/gpl-2.0.php GNU Public License
 * @since       1.0
 */

// Exit if accessed directly
if ( ! defined( 'ABSPATH' ) ) exit;

class CN_Certifications_Widget extends WP_Widget {

	public function __construct() {

		$options = array(
			'description' => __( 'Business Certifications', 'connections_certifications' )
		);

		parent::__construct(
			'cn_certifications',
			'Connections : ' . __( 'Business Certifications', 'connections_certifications' ),
			$options
		);
	}

	/**
	 * Registers the widget with the WordPress Widget API.
	 *
	 * @access public
	 * @since  1.0
	 *
	 * @return void
	 */
	public static function register() {

		register_widget( __CLASS__ );
	}

	/**
	 * Logic for handling updates from the widget form.
	 *
	 * @access  private
	 * @since  1.0
	 *
	 * @param array $new
	 * @param array $old
	 *
	 * @return array
	 */
	public function update( $new, $old ) {

		$new['title'] = strip_tags( $new['title'] );

		return $new;
	}

	/**
	 * Function for handling the widget control in admin panel.
	 *
	 * @access  private
	 * @since   1.0
	 *
	 * @param array $instance
	 *
	 * @return void
	 */
	public function form( $instance ) {

		// Setup the default widget options.
		$title = isset( $instance['title'] ) && strlen( $instance['title'] ) > 0 ? esc_html( $instance['title'] ) : __( 'Certifications', 'connections_certifications' );

		cnHTML::text(
			array(
				'prefix' => '',
				'class'  => 'widefat',
				'id'     => $this->get_field_id( 'title' ),
				'name'   => $this->get_field_name( 'title' ),
				'label'  => __( 'Title:', 'connections_certifications' ),
				'before' => '<p>',
				'after'  => '</p>',
			),
			$title
		);

	}

	/**
	 * Function for displaying the widget on the page.
	 *
	 * @access  private
	 * @since   1.0
	 *
	 * @param  array $args
	 * @param  array $option
	 */
	public function widget( $args, $option ) {

		// Only process and display the widget if displaying a single entry.
		if ( $slug =  cnQuery::getVar( 'cn-entry-slug' ) ) {

			// Grab an instance of the Connections object.
			//$instance = Connections_Directory();

			// Query the entry.
			$result = Connections_Directory()->retrieve->entry( urldecode( $slug ) );

			// Setup the entry object
			$entry = new cnEntry( $result );

			/**
			 * Extract $before_widget, $after_widget, $before_title and $after_title.
			 *
			 * @var $before_widget
			 * @var $after_widget
			 * @var $before_title
			 * @var $after_title
			 */
			extract( $args );

			// Setup the default widget options if they were not set when they were added to the sidebar;
			// ie. the user did not click the "Save" button on the widget.
			$title = strlen( $option['title'] ) > 0 ? $option['title'] : __( 'Certifications', 'connections_certifications' );

			// Setup the atts to be passed to the method that displays the data.
			$atts = array();

			echo $before_widget;

			echo $before_title . $title . $after_title;

			// Display the income level.
			Connections_Certifications::block( $entry, $atts, NULL );

			echo $after_widget;

		}
	}

}


connections-certifications.php
<?php
/**
 * An extension for the Connections Business Directory plugin which adds the ability to add and assign certifications to your business directory entries.
 *
 * @package   Connections Business Directory Extension - Certifications
 * @category  Extension
 * @author    Steven A. Zahm
 * @license   GPL-2.0+
 * @link      https://connections-pro.com
 * @copyright 2020 Steven A. Zahm
 *
 * @wordpress-plugin
 * Plugin Name:       Connections Business Directory Extension - Certifications
 * Plugin URI:        https://connections-pro.com/documentation/certifications/
 * Description:       An extension for the Connections Business Directory plugin which adds the ability to add and assign certifications to your business directory entries.
 * Version:           1.4
 * Author:            Steven A. Zahm
 * Author URI:        https://connections-pro.com
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Text Domain:       connections_certifications
 * Domain Path:       /languages
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

if ( ! class_exists( 'Connections_Certifications' ) ) {

	final class Connections_Certifications {

		const VERSION = '1.4';

		/**
		 * @var string The absolute path this this file.
		 *
		 * @access private
		 * @since 1.0
		 */
		private static $file = '';

		/**
		 * @var string The URL to the plugin's folder.
		 *
		 * @access private
		 * @since 1.0
		 */
		private static $url = '';

		/**
		 * @var string The absolute path to this plugin's folder.
		 *
		 * @access private
		 * @since 1.0
		 */
		private static $path = '';

		/**
		 * @var string The basename of the plugin.
		 *
		 * @access private
		 * @since 1.0
		 */
		private static $basename = '';

		public function __construct() {

			self::$file       = __FILE__;
			self::$url        = plugin_dir_url( self::$file );
			self::$path       = plugin_dir_path( self::$file );
			self::$basename   = plugin_basename( self::$file );

			self::loadDependencies();

			/**
			 * This should run on the `plugins_loaded` action hook. Since the extension loads on the
			 * `plugins_loaded` action hook, load immediately.
			 */
			cnText_Domain::register(
				'connections_certifications',
				self::$basename,
				'load'
			);

			// Add to Connections menu.
			add_filter( 'cn_submenu', array( __CLASS__, 'addMenu' ) );

			// Remove the "View" link from the "Certifications" taxonomy admin page.
			add_filter( 'cn_certification_row_actions', array( __CLASS__, 'removeViewAction' ) );

			// Register the metabox.
			add_action( 'cn_metabox', array( __CLASS__, 'registerMetabox') );

			// Attach certifications to entry when saving an entry.
			add_action( 'cn_process_taxonomy-category', array( __CLASS__, 'attachCertifications' ), 9, 2 );

			// Add support for CSV Import.
			add_filter( 'cncsv_map_import_fields', array( __CLASS__, 'import_field_option' ) );
			add_action( 'cncsv_import_fields', array( __CLASS__, 'import_field' ), 10, 3 );

			// Add support for CSV Export.
			add_filter( 'cn_csv_export_fields_config', array( __CLASS__, 'export_field_config' ) );
			add_filter( 'cn_csv_export_fields', array( __CLASS__, 'export_field_header' ) );
			add_filter( 'cn_export_header-certifications', array( __CLASS__, 'export_header' ), 10, 3 );
			add_filter( 'cn_export_field-certifications', array( __CLASS__, 'export_data' ), 10, 4 );

			// Add the "Certifications" option to the admin settings page.
			// This is also required so it'll be rendered by $entry->getContentBlock( 'certifications' ).
			add_filter( 'cn_content_blocks', array( __CLASS__, 'settingsOption') );

			// Add the action that'll be run when calling $entry->getContentBlock( 'certifications' ) from within a template.
			add_action( 'cn_entry_output_content-certifications', array( __CLASS__, 'block' ), 10, 3 );

			// Register the widget.
			add_action( 'widgets_init', array( 'CN_Certifications_Widget', 'register' ) );
		}

		/**
		 * The widget.
		 *
		 * @access private
		 * @since  1.0
		 * @static
		 * @return void
		 */
		private static function loadDependencies() {

			require_once( self::$path . 'includes/class.widgets.php' );
		}

		/**
		 * Register the taxonomy with the Gravity Forms Connector.
		 *
		 * @since 1.3
		 *
		 * @param array $taxonomy
		 *
		 * @return array
		 */
		public static function registerCertificationsTaxonomy( $taxonomy ) {

			$taxonomy['certification'] = array(
				'labels' => array(
					'name'          => _x(
						'Certifications',
						'Taxonomy field plural name.',
						'connections_gravity_forms'
					),
					'all_items'     => __( 'All Certifications', 'connections_gravity_forms' ),
					'select_items'  => __( 'Select Certifications', 'connections_gravity_forms' ),
					'singular_name' => _x(
						'Certification',
						'Taxonomy field singular name.',
						'connections_gravity_forms'
					),
					'field_label'   => __( 'Entry Certifications', 'connections_gravity_forms' ),
				),
			);

			return $taxonomy;
		}

		public static function addMenu( $menu ) {

			$menu[66]  = array(
				'hook'       => 'certifications',
				'page_title' => 'Connections : ' . __( 'Certifications', 'connections_certifications' ),
				'menu_title' => __( 'Certifications', 'connections_certifications' ),
				'capability' => 'connections_edit_categories',
				'menu_slug'  => 'connections_certifications',
				'function'   => array( __CLASS__, 'showPage' ),
			);

			return $menu;
		}

		public static function showPage() {

			// Grab an instance of the Connections object.
			$instance = Connections_Directory();

			if ( $instance->dbUpgrade ) {

				include_once CN_PATH . 'includes/inc.upgrade.php';
				connectionsShowUpgradePage();
				return;
			}

			switch ( $_GET['page'] ) {

				case 'connections_certifications':
					include_once self::$path . 'includes/admin/pages/certifications.php';
					connectionsShowCertificationsPage();
					break;
			}
		}

		public static function removeViewAction( $actions ) {

			unset( $actions['view'] );

			return $actions;
		}

		/**
		 * Registered the custom metabox.
		 *
		 * @access private
		 * @since  1.0
		 * @static
		 */
		public static function registerMetabox() {

			$atts = array(
				'id'       => 'certifications',
				'title'    => __( 'Certifications', 'connections_certifications' ),
				//'pages'    => $pages,
				'context'  => 'side',
				'priority' => 'core',
				'callback' => array( __CLASS__, 'metabox' ),
			);

			cnMetaboxAPI::add( $atts );
		}

		/**
		 * The certifications metabox.
		 *
		 * @access public
		 * @since  1.0
		 *
		 * @param  cnEntry $entry   An instance of the cnEntry object.
		 * @param  array   $metabox The metabox options array from self::register().
		 */
		public static function metabox( $entry, $metabox ) {

			echo '<div class="certificationdiv" id="taxonomy-certification">';

			$style = <<<HEREDOC
<style type="text/css" scoped>
	.certificationdiv div.tabs-panel {
		min-height: 42px;
		max-height: 200px;
		overflow: auto;
		padding: 0 0.9em;
		border: solid 1px #ddd;
		background-color: #fdfdfd;
	}
	.certificationdiv ul.certificationchecklist ul {
		margin-left: 18px;
	}
</style>
HEREDOC;
			echo $style;

				echo '<div id="certification-all" class="tabs-panel">';

				cnTemplatePart::walker(
					'term-checklist',
					array(
						'name'     => 'entry_certification',
						'taxonomy' => 'certification',
						'selected' => cnTerm::getRelationships( $entry->getID(), 'certification', array( 'fields' => 'ids' ) ),
					)
				);

				echo '</div>';
			echo '</div>';
		}

		/**
		 * Add, update or delete the entry certifications.
		 *
		 * @access public
		 * @since  1.0
		 * @static
		 *
		 * @param  string $action The action to being performed to an entry.
		 * @param  int    $id     The entry ID.
		 */
		public static function attachCertifications( $action, $id ) {

			// Grab an instance of the Connections object.
			$instance = Connections_Directory();

			if ( isset( $_POST['entry_certification'] ) && ! empty( $_POST['entry_certification'] ) ) {

				$instance->term->setTermRelationships( $id, $_POST['entry_certification'], 'certification' );

			} else {

				$instance->term->setTermRelationships( $id, array(), 'certification' );
			}
		}

		/**
		 * Add the field to the choices available to map to a CSV file field.
		 *
		 * @param array $fields
		 *
		 * @return array
		 */
		public static function import_field_option( $fields ) {

			$fields['certifications'] = 'Certifications';

			return $fields;
		}

		/**
		 * Import certifications and attach them to the entry.
		 *
		 * @param int         $id    The entry ID.
		 * @param array       $row   The parsed data from the CSV file.
		 * @param cnCSV_Entry $entry An instance of the cnCSV_Entry object.
		 */
		public static function import_field( $id, $row, $entry ) {

			$termIDs = array();

			$parsed = $entry->arrayPull( $row, 'certifications', $termIDs );

			if ( ! empty( $parsed ) ) {

				/*
				 * Convert the supplied certifications to an array and sanitize.
				 * Apply the same filters added to the core WP default filters for `pre_term_name` so the certification name
				 * will return a match if it exists.
				 */
				$certifications = explode( ',', $parsed );
				$certifications = array_map( 'sanitize_text_field', $certifications );
				$certifications = array_map( 'wp_filter_kses', $certifications );
				$certifications = array_map( '_wp_specialchars', $certifications );

				foreach ( $certifications as $certification ) {

					// Query the db for the term to be added.
					$term = cnTerm::getBy( 'name', $certification, 'certification' );

					if ( $term instanceof cnTerm_Object ) {

						$termIDs[] = $term->term_id;

					} else {

						// Add the new certification.
						$insert_result = cnTerm::insert( $certification, 'certification', array( 'slug' => '', 'parent' => '0', 'description' => '' ) );

						if ( ! is_wp_error( $insert_result ) ) {

							$termIDs[] = $insert_result['term_id'];
						}
					}
				}

			}

			// Do not set certification relationships if $termIDs is empty because if updating, it will delete existing relationships.
			if ( ! empty( $termIDs ) ) Connections_Directory()->term->setTermRelationships( $id, $termIDs, 'certification' );
		}

		/**
		 * Callback for the `cn_csv_export_fields_config` filter.
		 *
		 * Add the certifications export configurations option to the export config.
		 *
		 * @since 1.2
		 *
		 * @param array $fields
		 *
		 * @return array
		 */
		public static function export_field_config( $fields ) {

			$fields[] = array(
				'field'  => 'certifications',
				'type'   => 'certifications',
				//'fields' => '',
				//'table'  => CN_ENTRY_TABLE_META,
				//'types'  => NULL,
			);

			return $fields;
		}

		/**
		 * Callback for the `cn_csv_export_fields` filter.
		 *
		 * Set the column header name.
		 *
		 * @since 1.2
		 *
		 * @param array $fields
		 *
		 * @return array
		 */
		public static function export_field_header( $fields ) {

			$fields['certifications'] = 'Certifications';

			return $fields;
		}

		/**
		 * Callback for the `cn_export_header-certifications` filter.
		 *
		 * Returns the CSV file header name for the Certifications column.
		 *
		 * @since 1.2
		 *
		 * @param string                 $header
		 * @param array                  $atts
		 * @param cnCSV_Batch_Export_All $export
		 *
		 * @return string
		 */
		public static function export_header( $header, $atts, $export ) {

			return 'Certifications';
		}

		/**
		 * Callback for the `cn_export_field-certifications` filter.
		 *
		 * Export the data.
		 *
		 * @since 1.2
		 *
		 * @param string                 $data
		 * @param object                 $entry
		 * @param array                  $field
		 * @param cnCSV_Batch_Export_All $export
		 *
		 * @return string
		 */
		public static function export_data( $data, $entry, $field, $export ) {

			$data = '';

			// Process terms table and list all certifications in a single cell...
			$names = array();

			$terms = $export->getTerms( $entry->id, 'certification' );

			foreach ( $terms as $term ) {

				$names[] = $term->name;
			}

			if ( ! empty( $names ) ) {

				$data = $export->escapeAndQuote( implode( ',', $names ) );
			}

			return $data;
		}

		/**
		 * Add the custom meta as an option in the content block settings in the admin.
		 * This is required for the output to be rendered by $entry->getContentBlock().
		 *
		 * @access private
		 * @since  1.0
		 * @param  array  $blocks An associative array containing the registered content block settings options.
		 * @return array
		 */
		public static function settingsOption( $blocks ) {

			$blocks['certifications'] = __( 'Certifications', 'connections_certifications' );

			return $blocks;
		}

		/**
		 * Callback for the `cn_entry_output_content-{id}` filter in @see cnOutput::getCategoryBlock()
		 *
		 * Renders the Certifications content block.
		 * Modelled after the @see cnOutput::getCategoryBlock()
		 *
		 * @access  private
		 * @since   1.0
		 * @static
		 *
		 * @param  cnEntry    $object
		 * @param  array      $atts     The shortcode atts array passed from the calling action.
		 * @param  cnTemplate $template
		 */
		public static function block( $object, $atts, $template ) {

			global $wp_rewrite;

			$defaults = array(
				'container_tag'    => 'div',
				//'label_tag'        => 'span',
				'item_tag'         => 'span',
				'type'             => 'list',
				'list'             => 'unordered',
				//'label'            => __( 'Certifications:', 'connections_certifications' ) . ' ',
				'separator'        => ', ',
				'parent_separator' => ' &raquo; ',
				'before'           => '',
				'after'            => '',
				'link'             => FALSE,
				'parents'          => FALSE,
				//'child_of'         => 0,
				//'return'           => FALSE,
			);

			/**
			 * Allow extensions to filter the method default and supplied args.
			 *
			 * @since 1.0
			 */
			$atts = cnSanitize::args(
				apply_filters( 'cn_output_atts_certification', $atts ),
				apply_filters( 'cn_output_default_atts_certification', $defaults )
			);

			$terms = cnRetrieve::entryTerms( $object->getId(), 'certification' );

			if ( empty( $terms ) ) {

				return;
			}

			$count = count( $terms );
			$html  = '';
			$label = '';
			$items = array();

			if ( 'list' == $atts['type'] ) {

				$atts['item_tag'] = 'li';
			}

			$i = 1;

			foreach ( $terms as $term ) {

				$text = '';
				//$text .= esc_html( $term->name );

				if ( $atts['parents'] ) {

					// If the term is a root parent, skip.
					if ( 0 !== $term->parent ) {

						$text .= self::getTermParents(
							$term->parent,
							'certification',
							array(
								'link'       => $atts['link'],
								'separator'  => $atts['parent_separator'],
								'force_home' => $object->directoryHome['force_home'],
								'home_id'    => $object->directoryHome['page_id'],
							)
						);
					}
				}

				$atts['link'] = FALSE;
				if ( $atts['link'] ) {

					$rel = is_object( $wp_rewrite ) && $wp_rewrite->using_permalinks() ? 'rel="category tag"' : 'rel="category"';

					$url = cnTerm::permalink(
						$term,
						'category',
						array(
							'force_home' => $object->directoryHome['force_home'],
							'home_id'    => $object->directoryHome['page_id'],
						)
					);

					$text .= '<a href="' . $url . '" ' . $rel . '>' . esc_html( $term->name ) . '</a>';

				} else {

					$text .= esc_html( $term->name );
				}

				/**
				 * @since 1.0
				 */
				$items[] = apply_filters(
					'cn_entry_output_certification_item',
					sprintf(
						'<%1$s class="cn-certification-name cn-certification-%2$d">%3$s%4$s</%1$s>',
						$atts['item_tag'],
						$term->term_id,
						$text,
						$count > $i && 'list' !== $atts['type'] ? esc_html( $atts['separator'] ) : ''
					),
					$term,
					$count,
					$i,
					$atts
				);

				$i++; // Increment here so the correct value is passed to the filter.
			}

			/*
			 * Remove NULL, FALSE and empty strings (""), but leave values of 0 (zero).
			 * Filter our these in case someone hooks into the `cn_entry_output_category_item` filter and removes a category
			 * by returning an empty value.
			 */
			$items = array_filter( $items, 'strlen' );

			/**
			 * @since 1.0
			 */
			$items = apply_filters( 'cn_entry_output_certification_items', $items );

			if ( 'list' == $atts['type'] ) {

				$html .= sprintf(
					'<%1$s class="cn-certification-list">%2$s</%1$s>',
					'unordered' === $atts['list'] ? 'ul' : 'ol',
					implode( '', $items )
				);

			} else {

				$html .= implode( '', $items );
			}

			/**
			 * @since 1.0
			 */
			$html = apply_filters(
				'cn_entry_output_certification_container',
				sprintf(
					'<%1$s class="cn-certifications">%2$s</%1$s>' . PHP_EOL,
					$atts['container_tag'],
					$atts['before'] . $label . $html . $atts['after']
				),
				$atts
			);

			echo $html;
		}

		/**
		 * Retrieve category parents with separator.
		 *
		 * NOTE: This is the Connections equivalent of @see get_category_parents() in WordPress core ../wp-includes/category-template.php
		 *
		 * @access public
		 * @since  8.5.18
		 * @static
		 *
		 * @param int    $id        Term ID.
		 * @param string $taxonomy  Term taxonomy.
		 * @param array  $atts      The attributes array. {
		 *
		 *     @type bool   $link       Whether to format as link or as a string.
		 *                              Default: FALSE
		 *     @type string $separator  How to separate categories.
		 *                              Default: '/'
		 *     @type bool   $nicename   Whether to use nice name for display.
		 *                              Default: FALSE
		 *     @type array  $visited    Already linked to categories to prevent duplicates.
		 *                              Default: array()
		 *     @type bool   $force_home Default: FALSE
		 *     @type int    $home_id    Default: The page set as the directory home page.
		 * }
		 *
		 * @return string|WP_Error A list of category parents on success, WP_Error on failure.
		 */
		public static function getTermParents( $id, $taxonomy = 'certification', $atts = array() ) {

			$defaults = array(
				'link'       => FALSE,
				'separator'  => '/',
				'nicename'   => FALSE,
				'visited'    => array(),
				'force_home' => FALSE,
				'home_id'    => cnSettingsAPI::get( 'connections', 'connections_home_page', 'page_id' ),
			);

			$atts = cnSanitize::args( $atts, $defaults );

			$chain  = '';
			$parent = cnTerm::get( $id, $taxonomy );

			if ( is_wp_error( $parent ) ) {

				return $parent;
			}

			if ( $atts['nicename'] ) {

				$name = $parent->slug;

			} else {

				$name = $parent->name;
			}

			if ( $parent->parent && ( $parent->parent != $parent->term_id ) && ! in_array( $parent->parent,  $atts['visited'] ) ) {

				$atts['visited'][] = $parent->parent;

				$chain .= self::getTermParents( $parent->parent, 'certification', $atts );
			}

			if ( $atts['link'] ) {

				$chain .= '<span class="cn-category-breadcrumb-item" id="cn-category-breadcrumb-item-' . esc_attr( $parent->term_id ) . '">' . '<a href="' . esc_url( cnTerm::permalink( $parent->term_id, 'category', $atts ) ) . '">' . $name . '</a>' . $atts['separator'] . '</span>';

			} else {

				$chain .= $name . esc_html( $atts['separator'] );
			}

			return $chain;
		}
	}

	/**
	 * Start up the extension.
	 *
	 * @access public
	 * @since 1.0
	 *
	 * @return mixed object | bool
	 */
	function Connections_Certifications() {

			if ( class_exists('connectionsLoad') ) {

					return new Connections_Certifications();

			} else {

				add_action(
					'admin_notices',
					 create_function(
						 '',
						'echo \'<div id="message" class="error"><p><strong>ERROR:</strong> Connections must be installed and active in order use Connections Certifications.</p></div>\';'
						)
				);

				return FALSE;
			}
	}

	/**
	 * Since Connections loads at default priority 10, and this extension is dependent on Connections,
	 * we'll load with priority 11 so we know Connections will be loaded and ready first.
	 */
	add_action( 'plugins_loaded', 'Connections_Certifications', 11 );

	// Support the Gravity Form Connector, register Disciplines Taxonomy.
	// NOTE: Must add filter before the `plugins_loaded` action.
	add_filter( 'Connections_Directory\Connector\Gravity_Forms\Register_Taxonomy_Fields', array( 'Connections_Certifications', 'registerCertificationsTaxonomy' ) );

}
